### PR TITLE
LPS-159555  Add type declaration to ClassicEditor component in frontend-editor-ckeditor-web module

### PR DESCRIPTION
**Pull request by danvb112**
Related Issue: https://issues.liferay.com/browse/LPS-159555

### Motivation

Hello Reviewer, I'm working on a technical debt to remove some `@ts-ignore` annotations from the object code. For this I have to create some type declarations for components that were not developed in typescript, I can declare these types in the `object-js-components-web` module, but I think it will be restricted just for our case, so I decided to declare the types in the `frontend -editor-ckeditor-web` which is the root of the `ClassicEditor` component I had to remove the `@ts-ignore`, so anyone who wants to use this component with typescript in the future doesn't need to declare the same types in their own module, avoiding code repetition. If this idea is approved, I will send other PRs with new type declarations for other components we are using. What do you think?

### CODE REVIEW

**Comment by diegonvs** [comment]
> What do you think?

Thanks for working on this! I would wait @markocikos for a final comment but this PR looks ok

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+export interface IEditor {
+	editor: {
+		config: {contentsLangDirection: unknown; contentsLanguage: unknown};
+		setData: (data: unknown) => void;
+	};
+}
+
+interface IClassicEditorProps {
+	contents?: string;
+	editorConfig: string;
```
Comment (diegonvs) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts:31]: Hey @danvb112, editorConfig is an object

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+export interface IEditor {
+	editor: {
+		config: {contentsLangDirection: unknown; contentsLanguage: unknown};
+		setData: (data: unknown) => void;
+	};
+}
+
+interface IClassicEditorProps {
+	contents?: string;
```
Comment (diegonvs) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts:30]: I think `contents` is not optional.

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	editorConfig: string;
+	onChange: (content: string) => void;
+	ref: React.RefObject<IEditor>;
+}
```
Comment (diegonvs) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts:34]: Could you add missing `initialToolbarSet`, `name`, `title`? See https://cs.github.com/liferay/liferay-portal/blob/ca8a5112e6253e866f35987b8362a9305e174b51/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js#L23 for reference

Comment (danvb112) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts:37]: @diegonvs, Thanks for the review, I've added the changes you suggested :smile:

**Comment by danvb112** [comment]
ci:test:sf

**Comment by danvb112** [comment]
ci:test:relevant

**Comment by danvb112** [comment]
ci:test:relevant

**Comment by diegonvs** [comment]
ci:test:sf

**Comment by diegonvs** [comment]
ci:test:relevant

**Comment by diegonvs** [comment]
Now, config property is mirroring CKEditor's config intead of just an object :)

**Comment by danvb112** [comment]
ci:test:relevant

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	onChange,
+	ref,
+}: IClassicEditorProps): JSX.Element;
+
+export interface IEditor {
+	editor: CKEDITOR.editor;
+}
+
+interface IBalloonEditorProps {
+	config?: CKEDITOR.config;
```
Comment (markocikos) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts:35]: There are many places in this PR where the value of property interface is `CKEDITOR.config` or `CKEDITOR.editor`. This does not look right. Value should be a primitive, a `type` or another `interface`, right?

Comment (diegonvs) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts:35]: > CKEDITOR.config
In cases we're setting editor config

> CKEDITOR.editor
For cases where we're receving editor?

I'm not understanding you, could you bring more details?

Comment (markocikos) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts:35]: This is a TypeScript `interface`. `interface` determines object property _types_, not _values_. `string` is an example of type. `CKEDITOR.config` and `CKEDITOR.editor` are values. It's probably good enough to say `config: object` on this line, as those objects are too complex to drill down into.

I am not an expert on TS, so I'm not 100% sure if this is some shorthand way of setting types, but it _looks_ wrong. By the book, types of objects properties should be [primitives, a `type` or another `interface`](https://www.typescriptlang.org/docs/handbook/2/objects.html).

Comment (diegonvs) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts:35]: >  not values

but `config` in this case is another interface https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ckeditor4/index.d.ts#L585
image

and `editor` is a [namespace on definition](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ckeditor4/index.d.ts#L1081) and can be used to get more types for editor stuff

I'm not understanding you when you say values 😅 

>  It's probably good enough to say config: object on this line, as those objects are too complex to drill down into.

but how these types can be more reliable? I'll remove ckeditor types then and continue with the simpler object to not block this PR

> I am not an expert on TS

me neither, we are together 🤝🫂

**Comment by diegonvs** [comment]
ci:test:sf

**Comment by diegonvs** [comment]
ci:test:relevant

**Comment by diegonvs** [comment]
ci:test:relevant

**Comment by diegonvs** [comment]
ci:test:sf

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts (hunk 4 above)
Comment (markocikos) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/index.d.ts]: >config in this case is another interface

Oh, I see. Yeah, this is fine then :+1:

**Comment by diegonvs** [comment]
ci:forward

**Comment by diegonvs** [comment]
ci:test:relevant

**Comment by danvb112** [comment]
ci:test:relevant

**Comment by diegonvs** [comment]
Everything on this relevant was so flaky that I need to re-run ci again

**Comment by diegonvs** [comment]
ci:test:relevant

**Comment by mateusralv** [comment]
I couldn’t analyze the failures because [testray borken](https://liferay.slack.com/archives/C01MVBFRN0K/p1660308376284469) :/

**Comment by mateusralv** [comment]
ci:test:relevant

**Comment by danvb112** [comment]
ci:test:relevant

**Comment by mateusralv** [comment]
ci:test:relevant

**Comment by danvb112** [comment]
ci:test:relevant

**Comment by diegonvs** [comment]
ci:test:relevant

**Comment by susanchen3** [comment]
Taking a look now! :detective:

**Comment by susanchen3** [comment]
Reviewed, [tests failures](https://testray.liferay.com/home/-/testray/runs/compare?view=details&testrayRunIdA=1945892709&testrayRunIdB=1947261193&statusA=2&statusB=3&cur=1&delta=20&orderByCol=testrayComponentName&orderByType=asc) are unrelated :heavy_check_mark: Looks good to manually forward from QA perspective :)

**Comment by diegonvs** [comment]
https://github.com/liferay-frontend/liferay-portal/pull/2608

**Pull request closed without merging**